### DISCUSSION TITLE
[RFC] Tree support `titleExtra`

### DISCUSSION THREAD

**Discussion by zombieJ** [ant-design/ant-design#43259] [RFCs]
# 📋 RFC

## Summary

Tree provide `titleExtra` to add additional content injection.

## Basic example

```tsx
import { Tree } from `antd`;

render(
  <Tree
    treeData={[{ title: 'root', key: 'root' }]}
    titleExtra={(nodeData) => nodeData.title}
  />
);
```

## Motivation

Tree with additional node operation is common usage in real world. Developer have to patch extra effect to realize this. It's better to provided by default.

## Detailed

`titleExtra` accept nodeData as param for customize:

```tsx
type titleExtra = (data: TreeNodeData) => React.ReactNode;
```

And add node on the end of edge of node title. Which should not override title content.

## Drawbacks

Noop

## Adoption strategy

Provided in feature branch.

**Comment by kiner-tang** [comment]
是不是可以把选中的节点信息也传过来，很多场景需要获取当前选中节点的信息

**Reply by Wxh16144** [reply]
选中的也暴露出来感觉太冗余了, 完全可以通过受控的形式得到已选择的节点信息

**Reply by zombieJ** [reply]
选中信息 HOC 的话都可以拿到

**Comment by poyiding** [comment]
加上 titleExtra 后每个节点都会渲染， 需要考虑怎么方便判断 某个节点不需要titleExtra 。

**Reply by kiner-tang** [reply]
这个 titleExtra 是不是说的是整个 tree 的 title，而不是每一个节点的 title 呀

**Reply by kiner-tang** [reply]
如果要每个节点加的话，感觉节点那边也可以加一个 titleExtra 属性

**Reply by poyiding** [reply]
我理解的是titleExtra 会对每个节点的title 都注入，所以 需要考虑不需要的时候怎么方便判断

**Reply by RedJue** [reply]
感觉应该都需要加，整体和具体节点属性，默认继承整体、也可以具体节点修改，节点权限场景还是挺常见的。

**Comment by mumiao** [comment]
有诉求 +1

**Comment by zombieJ** [comment]
https://codesandbox.io/s/ji-ben-antd-5-6-4-forked-rmhwhh?file=/demo.tsx

使用 `titleRender` 后可以通过 Flex 布局实现，从 API 角度看有了会更方便，但是目前也能实现。优先级靠后。
